parser_library/test/processing/end_instr_test.cpp

```diff
@@ -0,0 +1,320 @@
+TEST CSECT                                                                                 
+     MAC  
+     GIBBERISH
+)");
+    analyzer a(input);
+    a.analyze();
+    a.collect_diags();
+    EXPECT_TRUE(a.diags().empty());
+}
+// todo , warning for statements after end not implemented
+TEST(END, end_called_from_macro_with_look_ahead_symbols_found)
+{
+    std::string input(R"( 
+         MACRO                                               
+     MAC                                                 
+     END   TEST                                          
+     MEND                                                
+TEST CSECT                                                   
+     AIF   (L'X LT 0).X                                 
+     MAC
+     BR    14                                            
+.X   ANOP  ,                                            
+X    DS    C                                                          
+     BR    0                                             
+     END                                                 
+)");
+    analyzer a(input);
+    a.analyze();
+    a.collect_diags();
+    EXPECT_TRUE(a.diags().empty());
+}
+
+TEST(END, end_called_from_macro_with_look_ahead_symbols_not_found)
+{
+    std::string input(R"( 
+         MACRO                                               
+     MAC                                                 
+     END   TEST                                          
+     MEND                                                
+TEST CSECT                                                   
+     AIF   (L'X LT 0).X                                 
+     MAC
+     BR    14                                                                                                    
+     BR    0                                             
+     END                                                 
+)");
+    analyzer a(input);
+    a.analyze();
+    a.collect_diags();
+    EXPECT_TRUE(matches_message_codes(a.diags(), { "W013" }));
+}
+
+TEST(END, end_called_from_copybook_inside_macro_one)
+{
+    std::string copybook_content = R"(
+   MACRO
+  M
+  LR 1,2
+  END 
+  undef_opcode
+ MEND       
+)";
+    std::string input = R"(
+         COPY COPYBOOK_TWO
+         M
+        undef_opcode
+)";
+    mock_parse_lib_provider lib_provider { { "COPYBOOK_TWO", copybook_content } };
+    analyzer a(input, analyzer_options { &lib_provider });
+
+    a.analyze();
+    a.collect_diags();
+
+    EXPECT_TRUE(a.diags().empty());
+}
+TEST(END, end_called_from_copybook_)
+{
+    std::string copybook_content = R"(
+
+  LR 1,2
+  END       
+)";
+    std::string input = R"(
+         COPY COPYBOOK_TWO
+        undef_opcode
+)";
+    mock_parse_lib_provider lib_provider { { "COPYBOOK_TWO", copybook_content } };
+    analyzer a(input, analyzer_options { &lib_provider });
+
+    a.analyze();
+    a.collect_diags();
+
+    EXPECT_TRUE(a.diags().empty());
+}
+TEST(END, end_called_from_copybook_inside_macro_two)
```

Comment:
The way I understand the title, I'd expect COPY instruction inside the macro definition (which is a case I don't see covered so far).